conwayste/src/ui/context.rs

```diff
@@ -0,0 +1,216 @@
+pub enum EventType {
+    Click,
+    KeyPress,
+    Move, // mouse move
+    Translate,
+    Resize,
+    ParentTranslate,
+    ParentResize,
+    // TODO: not sure about Child* because we'd need a widget ID to say which child
+    //ChildTranslate,
+    //ChildResize,
+}
+
+// TODO: move this elsewhere; it's in here to keep separate from other code (avoid merge conflicts)
+#[derive(Debug, Clone)]
+pub struct Event {
+    pub what: EventType,
+    pub x: f32,  // usually x position, but width if resize event
+    pub y: f32,
+}
+
+#[allow(unused)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Handled {
+    Handled,  // no other handlers should be called
+    NotHandled, // continue calling handlers
+}
+
+pub type Handler = Box<dyn FnMut(&mut dyn EmitEvent, &mut UIContext, &Event) -> Result<Handled, Box<dyn Error>> + Send>;
+
+pub type HandlerMap = HashMap<EventType, Vec<Handler>>;
+
+/// Trait for widgets that can handle various events. Use `.on` to register a handler and `.emit`
+/// to emit an event which will cause all handlers for the event's type to be called.
+///
+/// # Errors
+///
+/// * It is an error to call `.emit` or `.on` from within a handler.
+pub trait EmitEvent: Downcast {
+    /// Setup a handler for an event type
+    ///
+    /// ```
+    /// let handler = |obj: &mut dyn EmitEvent, uictx: &mut context::UIContext, evt: &context::Event| {
+    ///     use context::Handled::*;
+    ///     let mut widget = obj.downcast_mut::<MyWidget>().unwrap();
+    ///
+    ///     //... do stuff
+    ///
+    ///     Ok(Handled) // can also return NotHandled to allow other handlers for this event type to run
+    /// };
+    /// my_widget.on(context::EventType::Click, Box::new(handler));
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// * It is an error to call this from within a handler.
+    fn on(&mut self, what: EventType, f: Handler) -> Result<(), Box<dyn Error>>;
+
+    /// Emit an event -- call all handlers for this event's type (as long as they return NotHandled)
+    ///
+    /// # Errors
+    ///
+    /// * It is an error to call this from within a handler.
+    /// * The first error to be returned by a handler will be returned here, and no other handlers
+    ///   will run.
+    fn emit(&mut self, event: &Event, uictx: &mut UIContext) -> Result<(), Box<dyn Error>>;
+}
+
+impl_downcast!(EmitEvent);
+
+/// Implement EmitEvent for a widget (though strictly speaking non-widgets can implement it).
+///
+/// # Example
+///
+/// ```
+/// struct MyWidget {
+///     handlers: Option<HandlerMap>,
+///     ...
+/// }
+///
+/// impl MyWidget {
+///     fn new() -> Self {
+///         MyWidget {
+///             handlers: Some(context::HandlerMap::new()),
+///             ...
+///         }
+///     }
+/// }
+/// // top level of the module
+/// impl_emit_event!(MyWidget, self.handlers);
+/// ```
+#[macro_export]
+macro_rules! impl_emit_event {
+    ($widget_name:ty, self.$handler_field:ident) => {
```

Comment:
Huh cool, didn't know you could do `self.$...`

Reply:
In a macro it's just a token tree. As long as you balance ([{, anything goes.